File: question2/src/main/java/com/example/question2/MainActivity.java
//Observe Lifecycle of Fragment with activity while using Add, Replace, hide, show, remove

package com.example.question2;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import android.os.Bundle;
import android.util.Log;
import android.view.View;

import java.util.Objects;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";

    FragmentManager fragmentManager;
    FragmentTransaction fragmentTransaction;
    Myfragment myfragment1;
    Myfragment2 myfragment2;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Log.d(TAG, "onCreate");

        myfragment1 = new Myfragment();
        myfragment2 = new Myfragment2();
    }


    public void ADD(View view) {
        if(!myfragment1.isAdded()){
            fragmentManager = getSupportFragmentManager();
            fragmentTransaction = fragmentManager.beginTransaction();
            fragmentTransaction.add(R.id.fragment, myfragment1, "Fragment 1");
            fragmentTransaction.addToBackStack(null);
            fragmentTransaction.commit();
        }
    }

    public void REPLACE(View view) {
        fragmentManager = getSupportFragmentManager();
        fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.replace(R.id.fragment, myfragment2, "Fragment 2");
        fragmentTransaction.addToBackStack(null);
        fragmentTransaction.commit();
    }

    public void HIDE(View view) {
        fragmentManager = getSupportFragmentManager();
        fragmentTransaction = fragmentManager.beginTransaction();
        if(myfragment1.isAdded()){
            fragmentTransaction.hide(myfragment1);
            fragmentTransaction.commit();
        }
        else{
            fragmentTransaction.hide(myfragment2);
            fragmentTransaction.commit();
        }


    }

    public void SHOW(View view) {
        fragmentManager = getSupportFragmentManager();
        fragmentTransaction = fragmentManager.beginTransaction();
        if(myfragment1.isAdded()){
            fragmentTransaction.show(myfragment1);
            fragmentTransaction.commit();
        }
        else{
            fragmentTransaction.show(myfragment2);
            fragmentTransaction.commit();
        }

    }

    public void REMOVE(View view) {
        fragmentManager = getSupportFragmentManager();
        fragmentManager.popBackStack();
    }


    @Override
    protected void onStart() {
        super.onStart();
        Log.d(TAG, "onStart");
    }

    @Override
    protected void onPause() {
        super.onPause();
        Log.d(TAG, "onPause");
    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.d(TAG, "onResume");
    }

    @Override
    protected void onRestart() {
        super.onRestart();
        Log.d(TAG, "onRestart");
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.d(TAG, "onStop");
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.d(TAG, "onDestroy");
    }
}

File: exercise11androidintegrationwithwebservices/src/main/java/com/example/exercise11androidintegrationwithwebservices/MainActivity.kt
/**
    1. Set background image of any layout. (via Glide and via HTTP URL connection)
        https://image.freepik.com/free-photo/image-human-brain_99433-298.jpg

    2. Create a list by parsing below url, using both Retrofit & HttpURLConnection
        https://storage.googleapis.com/network-security-conf-codelab.appspot.com/v2/posts.json
 */

package com.example.exercise11androidintegrationwithwebservices

import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.bumptech.glide.Glide
import com.bumptech.glide.request.target.SimpleTarget
import com.google.gson.Gson
import kotlinx.android.synthetic.main.activity_main.*
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.indeterminateProgressDialog
import org.jetbrains.anko.uiThread
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.BufferedInputStream
import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL


class MainActivity : AppCompatActivity() {
    private companion object {
        private const val TAG = "MainActivity"
    }

    private val gson = Gson()

    private val imageUrl = "https://image.freepik.com/free-photo/image-human-brain_99433-298.jpg"
    private val listUrl =
        "https://storage.googleapis.com/network-security-conf-codelab.appspot.com/v2/posts.json"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    /**
     * function to update the UI after downloading the image using Glide or Http Api
     */
    private fun updateUI(resource: Drawable) {
        Toast.makeText(applicationContext, "Downloaded", Toast.LENGTH_LONG).show()

        layout.background = resource
        textView.setTextColor(Color.WHITE)
        textView2.setTextColor(Color.WHITE)
    }

    /**
     * function to download the image using Glide
     */
    fun downloadImageGlide(view: View) {
        Glide.with(this).load(imageUrl)
            .into(object : SimpleTarget<Drawable?>() {
                override fun onResourceReady(
                    resource: Drawable,
                    transition: com.bumptech.glide.request.transition.Transition<in Drawable?>?
                ) {
                    updateUI(resource)
                }
            })
    }

    /**
     * function to download the image using Http Api
     */
    fun downloadImageHttpApi(view: View) {
        val dialog = indeterminateProgressDialog("Downloading")
        doAsync {
            dialog.show()
            println("Downloading");
            val url = URL(imageUrl);

            val urlConnection = url.openConnection() as HttpURLConnection

            try {
                val input: InputStream = BufferedInputStream(urlConnection.inputStream);
                val bitmap = BitmapFactory.decodeStream(input)
                val drawable = BitmapDrawable(resources, bitmap)

                uiThread {
                    updateUI(drawable)
                }

            } catch (e: IOException) {
                Log.e(TAG, e.toString())
            } finally {
                urlConnection.disconnect()
                dialog.dismiss()
            }
        }
    }

    /**
     * function to download the Json and Parse the Json object using Gson
     * and send the Parsable "MyData" class object to RecycleView class
     */
    fun downloadListUsingHttpApi(view: View) {
        val dialog = indeterminateProgressDialog("Downloading")
        doAsync {
            dialog.show()

            val json = URL(listUrl).readText()

            try {
                val data = gson.fromJson(json, MyData::class.java)
                val intent = Intent(applicationContext, RecyclerViewClass::class.java)
                intent.putExtra(RecyclerViewClass.MYDATA, data)
                startActivity(intent)

            } catch (e: IOException) {
                Log.e(TAG, e.toString())
            } finally {
                dialog.dismiss()
            }
        }
    }

    /**
     * function to download the Json using RetroFit Api
     * and send the Parsable "MyData" class object to RecycleView class
     */
    fun downloadListUsingRetroFit(view: View) {
        val call: Call<MyData> = ApiClient.getClient.getData()
        call.enqueue(object : Callback<MyData> {
            override fun onResponse(call: Call<MyData>?, response: Response<MyData>?) {

                val data: MyData = response!!.body()!!

                val intent = Intent(applicationContext, RecyclerViewClass::class.java)
                intent.putExtra(RecyclerViewClass.MYDATA, data)

                startActivity(intent)
            }

            override fun onFailure(call: Call<MyData>?, t: Throwable?) {
                Log.e(TAG, "Exception: " + t.toString())
            }
        })
    }
}

